### PR TITLE
RFC: Version Numbering

### LINKED ISSUE #2335: [Feat] Let the main branch have x.y.0 version
**Is your feature request related to a problem? Please describe.**

The version on the release branches always includes the third digit (aka micro or patch), so development versions are then 8.0.1dev.

However, the main branch with the development-only odd minor version numbers never gets a micro, e.g., 8.1dev, version which is inconsistent.

The missing third number also open room for inconsistencies in representing the version representation. It is 8.3dev or 8.3.dev?

**Describe the solution you'd like**

The main branch should add 0 to the version, i.e., 8.3.0dev. The 0 (and dev) will be always there.

**Describe alternatives you've considered**

1. Keep the current state, but be consistent in 8.3dev versus 8.3.dev. (Not realistic and does not address inconsistencies with release branches.)
2. Replace dev by 0, i.e., 8.3.0. The odd minor says that it is a development version. (In odds common versioning schemes which add some type of suffix.) 
3. Drop the version idea completely from main and somehow have it version-less with commit has used everywhere. (Too difficult to implement.)
4. The main branch has the next version to be branched from it, so now it would be 8.3.0dev or 8.4.0dev. This includes either not using the odd-even distinction and threading odd minor versions (main is 8.3.0dev) the same as even ones or skipping odd minor versions completely (main is 8.4.0dev) to keep the practice just for releases. (Although this seems like a big change, it is only change in procedure and it actually greatly simplifies the numbering procedure. It may also allow us to use some existing tooling to automate things and, e.g., improve the suffixes.)
5. Release previews from the main branch, so keep on main whatever is the next version to be released from that code base, but keeping the odd-even distinction. The micro version than increases as needed. (The would need an easy release procedure to be feasible, but releasing unstable previews directly from main is not incompatible with the preferred solution of just including 0 right now.)

**Additional context**

We had several consistency issues recently and the missing micro release complicates the procedure (now captured in a script in #2331).

This issue is partially based on my previous comment: https://github.com/OSGeo/grass/issues/1583#issuecomment-1004982301.

### PR DESCRIPTION
**Pull request by wenzeslaus**
This PR is a RFC for a new version numbering system. The main point is not making any distinction between even and odd version numbers. After 8.2.z, 8.3.0 would follow.

Notably, this RFC does not follow the current RFC procedure (if any), but creates a PR as suggested in the past as a possible improvement of the procedure.

Closes #2335.

### CODE REVIEW

File: doc/development/rfc/version_numbering.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+## Background
+
+The increasing versions was always following the same ideas as in
+the Semantic Versioning, i.e., _major_, _minor_, and _patch_
+with major for breaking changes, minor for backwards compatible
+features, and patch for fixes.
+
+For the versions 7.0 till 8.2, there was a distinction between even and odd minor versions.
+Even minor versions were released, while odd minor version marked development version.
+This odd versions never had a patch number assigned and were never released.
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:27]: Note: The history to this was the adoption of the Linux kernel versioning system:

https://en.wikipedia.org/wiki/Software_versioning#Odd-numbered_versions_for_development_releases

Indeed, the Linux kernel project abandoned it time ago.

Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md:27]: Any idea when it was first applied to GRASS GIS? Just to have the history complete there.

Also interesting: Linux kernel moved from 2 to 3 _not_ because of a breaking updates. What are the rules now, I don't know.

Comment (neteler) [reply doc/development/rfc/version_numbering.md:27]: Here a relevant snippet (not fully history):
https://lists.osgeo.org/pipermail/grass-dev/2007-August/032705.html

File: doc/development/rfc/version_numbering.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+This in turn helps to address the issue of outdated tutorials and small, but breaking, changes.
+
+Some major features such as changes in the GUI are fully backwards compatible
+(GUI API is not a public API in versions 7 and 8), but are heavily influencing tutorials and
+other teaching materials. In that sense, behavior linked to a particular interface is changed,
+so a major version is required.
+
+There is always a list of many small changes which are not backwards compatible,
+for example, a cleanup of depreciated functions. None of these changes alone does
+not seem worth a major release, but because it would require one, it is never done.
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:75]:
```suggestion
for example, a cleanup of depreciated functions. None of these changes alone 
seems worth a major release, but because it would require one, it is never done.
```

File: doc/development/rfc/version_numbering.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+Minor version increment is strongly recommend for all new functionality or improvements
+which are not bug fixes.
+
+### Micro
+
+If only backwards compatible bug fixes, i.e., fixes of incorrect behavior, are applied,
+a micro version can be released.
+
+The micro version is also known as _patch_ which is what Semantic Versioning is using
+and _point_ which is what was used in some GRASS GIS documents. The work _micro_ is used
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:91]:
```suggestion
and _point_ which is what was used in some GRASS GIS documents. The word _micro_ is used
```

File: doc/development/rfc/version_numbering.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+to avoid collision with patch in the meaning of individual changeset or fix
+(release contain one or more of these changes). Point release (or _dot_ release)
+can generally apply to anything after the first dot, so minor or micro release.
+
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:95]:
```suggestion
can generally apply to anything after the first dot, so minor or micro release
is preferred.

```

File: doc/development/rfc/version_numbering.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+Release candidates (RCs) are pre-releases marked by appending RC and
+a release candidate number to the version, e.g., 3.5.1RC2.
+
+Transition to the Semantic Versioning style for pre-releases is strongly recommend.
+Version number is followed by a dash (hyphen) followed by a dot-separated identifier
+which consists of identification of pre-release type, i.e., rc or RC, optional dot,
+and a release candidate number, e.g., 3.5.1-rc.2, 3.5.1-RC.2, 3.5.1-rc2, or 3.5.1-RC2.
+
+Daily builds and other builds of development versions other than release candidates
+are may use additional dashes to specify the actual source version, e.g.,
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:116]:
```suggestion
may use additional dashes to specify the actual source version, e.g.,
```

File: doc/development/rfc/version_numbering.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+won't parse the version correctly. A suggested workaround is to use the dash
+only in a file name and elsewhere use plus.
+
+Any version parsing or build system, must support the form with dash
+and the form with plus.
+
+## Relation to Other Documents
+
+* [RFC 4: Release Procedure](https://trac.osgeo.org/grass/wiki/RFC/4_ReleaseProcedure): No changes to release procedure, only the numbering is changed.
+* [Release Schedule](https://trac.osgeo.org/grass/wiki/Release/Schedule#Externalreleaseschedules) (from  at Trac wiki under Release): The Release Schedule document describes schedule, branching, release maintenance, and numbering. The numbering is changed, specifically the use of odd version numbers for development.
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:137]:
```suggestion
* [Release Schedule](https://trac.osgeo.org/grass/wiki/Release/Schedule#Externalreleaseschedules) (from Trac wiki under Release): The Release Schedule document describes schedule, branching, release maintenance, and numbering. The numbering is changed, specifically the use of odd version numbers for development.
```

File: doc/development/rfc/version_numbering.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+## Background
+
+The increasing versions was always following the same ideas as in
+the Semantic Versioning, i.e., _major_, _minor_, and _patch_
+with major for breaking changes, minor for backwards compatible
+features, and patch for fixes.
+
+For the versions 7.0 till 8.2, there was a distinction between even and odd minor versions.
+Even minor versions were released, while odd minor version marked development version.
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:26]:
```suggestion
Even minor versions were released, while odd minor versions marked development versions.
```

File: doc/development/rfc/version_numbering.md (hunk 1 above)
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:27]:
```suggestion
These odd versions never had a patch number assigned and were never released.
```

File: doc/development/rfc/version_numbering.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+## Motivation
+
+Explaining the version numbering should be as easy as possible.
+Ideally, it would not need any explanation at all.
+When all version numbers refer to releases, previews (alphas, betas, daily builds),
+always need additional indication of the version being a preview.
+However, using odd version to mark the development versions requires
+still additional explanation, for example, download for 8.1 on the website
+said _preview_ anyway because the odd number does not indicate the
+development version.
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:38]:
```suggestion
said _preview_ anyway because the odd number does not indicate a
development version by itself, i.e., it's not self-explained.
```

File: doc/development/rfc/version_numbering.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+Hence, odd numbers for development versions do not bring any advantage.
+
+The branching and release procedure with even minor numbers for releases
+and odd numbers for development versions require that when new branch for
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:42]:
```suggestion
and odd numbers for development versions require that when a new branch for
```

File: doc/development/rfc/version_numbering.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+both interface and behavior changes (API also defines what happens,
+not just names and signatures).
+
+Additionally, major feature additions which would require only minor release in terms of API stability,
+are strongly suggested to trigger major release as well.
+This in turn helps to address the issue of outdated tutorials and small, but breaking, changes.
+
+Some major features such as changes in the GUI are fully backwards compatible
+(GUI API is not a public API in versions 7 and 8), but are heavily influencing tutorials and
+other teaching materials. In that sense, behavior linked to a particular interface is changed,
```
Comment (veroandreo) [inline doc/development/rfc/version_numbering.md:70]:
```suggestion
other teaching materials. In that sense, the behavior linked to a particular interface is changed,
```

File: doc/development/rfc/version_numbering.md (hunk 4 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: The original wording was unclear. I added more explanation there.

File: doc/development/rfc/version_numbering.md (hunk 1 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: Thanks. Great material. I updated the document based on these two although I don't reference them in the text.

File: doc/development/rfc/version_numbering.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+There is always a list of many small changes which are not backwards compatible,
+for example, a cleanup of depreciated functions. None of these changes alone 
+seems worth a major release, but because it would require one, it is never done.
+With more common major releases, small changes can happen more often.
+
+### Minor
+
+Minor version must be incremented if new functionality is added
+or if existing functionality is marked as deprecated.
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:92]: Just nitpicking: deprecated or depreciated as above?

File: doc/development/rfc/version_numbering.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+- _micro_ is for fixes.
+These numbers are assigned to releases. Development versions, pre-release,
+and builds are marked by additional labels as suffixes.
+All major and minor version numbers are used for releases,
+i.e., there is no distinction between even and odd numbers.
+
+## Background
+
+The increasing versions was always following the same ideas as in
+the Semantic Versioning, i.e., _major_, _minor_, and _patch_
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:21]:
```suggestion
the [Semantic Versioning](https://semver.org/), i.e., _major_, _minor_, and _patch_
```

Link added as a guide for newcomers

File: doc/development/rfc/version_numbering.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+Even minor versions were released, while odd minor versions marked development versions.
+These odd versions never had a patch number assigned and were never released.
+
+The practice of odd minor version denoting development versions and even minor versions
+denoting releases was introduced with 5.0.0 and was followed in various ways in the
+version 5 and 6 series.
+This odd-even practice followed the numbering scheme of the Linux kernel
+which abandoned the practice since then.
+
+The odd-even practice replaced a system where multiple numbered beta
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:36]:
```suggestion
Historically, the odd-even practice replaced a system where multiple numbered beta
```

File: doc/development/rfc/version_numbering.md [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+and _point_ which is what was used in some GRASS GIS documents. The word _micro_ is used
+to avoid collision with patch in the meaning of individual changeset or fix
+(release contain one or more of these changes). Point release (or _dot_ release)
+can generally apply to anything after the first dot, i.e., minor or micro release.
+Using minor and micro does not have that ambiguity.
+Minor and micro have potential for confusion between these two due to the
+similarity of the names, but minor is an established term in this context
+and minor is sometimes used in this context and other contexts it has
+the right connotations.
+### Development Versions
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:111]:
```suggestion
the right connotations.

### Development Versions
```

File: doc/development/rfc/version_numbering.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+The micro version is also known as _patch_ which is what Semantic Versioning is using
+and _point_ which is what was used in some GRASS GIS documents. The word _micro_ is used
+to avoid collision with patch in the meaning of individual changeset or fix
+(release contain one or more of these changes). Point release (or _dot_ release)
+can generally apply to anything after the first dot, i.e., minor or micro release.
+Using minor and micro does not have that ambiguity.
+Minor and micro have potential for confusion between these two due to the
+similarity of the names, but minor is an established term in this context
+and minor is sometimes used in this context and other contexts it has
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:109]:
```suggestion
and minor is sometimes used in this context and other contexts it has
```

This sentence isn't entirely clear to me. Maybe "... while ..." ?

File: doc/development/rfc/version_numbering.md [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+3.5.1-dev-05e5df2e7.
+
+Any version parsing or build system, must support both systems, the one without
+a dash and the one with dash.
+
+### Build Information
+
+When build information is captured in the version number, Semantic Versioning
+prescribed plus sign as a separator, e.g., 3.5.1+1 or 3.5.1-RC1+1.
+However, filenames with a plus sign my not work well, so a dash as a separator
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:141]:
```suggestion
However, filenames with a plus sign may not work well, so a dash as a separator
```

File: doc/development/rfc/version_numbering.md (hunk 13 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: I agree that the fact that this is talking about history needs to be emphasizes here. I made it explicit.

```
[-The-]{+At the time of version 5, the+} odd-even practice replaced a system where multiple numbered beta
versions were released. Version 5.0beta10 was a last beta release before the first
pre-release of 5.0.0.
```

File: doc/development/rfc/version_numbering.md (hunk 12 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: Done for the first mention of Semantic Versioning. It is also linked in the related documents section.

```
The increasing versions was always following the same ideas as in
the [-Semantic Versioning,-]{+[Semantic Versioning](https://semver.org/),+}
i.e., _major_, _minor_, and [-_patch_-]{+_patch_,+}
with major for breaking changes, minor for backwards compatible
features, and patch for fixes.
```

File: doc/development/rfc/version_numbering.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: Hm, introduced when using the _Commit suggestion_ feature online; I did a bigger edit using that.

File: doc/development/rfc/version_numbering.md (hunk 15 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: I improved the whole paragraph. The last sentence about confusing minor and micro might have been confusing also because I confused minor and micro.

```
The micro version is also known as _patch_ which is what Semantic Versioning is using
and _point_ which is what was used in some GRASS GIS documents.
The word [-_micro_-]{+_micro_, rather than patch,+} is used
to avoid collision with patch [-in the meaning of-]{+referring to an+} individual changeset or fix
[-(release contain-]{+(a release contains+} one or more of these changes).
[-Point-]{+A point+} release (or {+a+} _dot_ release) can generally apply to anything after
the first dot, i.e., minor or micro [-release.-]
[-Using-]{+release, while+}
{+using+} minor and micro does not have that ambiguity.
[-Minor-]{+Although minor+} and micro have potential for confusion[-between these two-] due to the
similarity of [-the-]{+their+} names,[-but-] minor is an established term in this context
and [-minor-]{+micro+} is sometimes used in this context and {+in+} other [-contexts-]{+contexts,+}
it has the right connotations.
```

File: doc/development/rfc/version_numbering.md (hunk 16 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: Fixed.

**Comment by wenzeslaus** [comment]
Thanks @neteler for the comments. The minor versus micro is clearly an issue.

**Review by neteler — APPROVED** [review]
IMHO ready for PSC voting which should happen before the upcoming 8.3.0 release.

**Comment by wenzeslaus** [comment]
I identified some todos, but perhaps they are for next version of the RFC document:

* [ ] discourage 82 and 820 versions, recommend 8.2 and 8.2.0
* [ ] include recommendation for dates
* [ ] specifically discourage linking name and version
  * Example: don't use "GRASS GIS 8 includes foo and bar." when you simply mean "the most recent (current) version includes."
  * Example bad: "The best version was GRASS GIS 7." Example good: "The best version of GRASS GIS was 7."
  * "winGRASS 8.2.0/standalone" in release announcement of 8.2.0 does not make much sense. Obviously, the Windows binary is for the corresponding version.  
  *  In short comments: G8 is unclear, when the version is important, using v8 is common practice. 
  * If version is really needed "The default database driver used by GRASS GIS 7 is SQLite." (as opposed to v6), say "The default database driver used by GRASS GIS is SQLite (since version 7)." so that the version number is always valid - the version of introduction stays the same and does not change.
* [ ] consider clarifying that soft freeze is the branching (aka feature freeze) while code freeze is release RFC version (this came up during PSC meeting)

**Comment by wenzeslaus** [comment]
Perhaps discussion in #2575 is an opportunity to reflect again on what versioning and branching scheme we want to adopt. Perhaps what is now major.minor is just major, i.e., branch is major, not major.minor, and we drop micro.

**Comment by BadAssassin** [comment]
The micro number has not been terribly relevant for some time. GRASS 
generally does not do bug fix releases and address version at the same time.

I do like using odd minor numbers for development, but I think we could 
release with major.minor and kill micro. I'm not sure just using major 
would provide enough granularity between stable and unstable releases.

On 10/20/2022 6:01 PM, Vaclav Petras wrote:
>
> Perhaps discussion in #2575 
> <https://github.com/OSGeo/grass/issues/2575> is an opportunity to 
> reflect again on what versioning and branching scheme we want to 
> adopt. Perhaps what is now major.minor is just major, i.e., branch is 
> major, not major.minor, and we drop micro.
>
> Message ID: ***@***.***>
>
-- 
Best Regards,
-Brad

File: doc/development/rfc/version_numbering.md [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+Author: Vaclav Petras
+
+Status: Draft
+
+## Summary
+
+Use version number in a format _major.minor.micro_ where
+- _major_ is for large additions or backwards incompatible changes,
+- _minor_ is for additions, and
+- _micro_ is for fixes.
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:12]: I wonder if we 'd better follow SemVer with "patch" rather than "micro"?

File: doc/development/rfc/version_numbering.md [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+
+Any version parsing or build system, must support the form with dash
+and the form with plus.
+
+While the assignment of a version is typically done by tagging in a version
+control system, but builds are managed separately.
+
+## Relation to Other Documents
+
+* [RFC 4: Release Procedure](https://trac.osgeo.org/grass/wiki/RFC/4_ReleaseProcedure): No changes to release procedure, only the numbering is changed.
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:169]: (URL to be updated after #2815)

File: doc/development/rfc/version_numbering.md [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+* [Release Schedule](https://trac.osgeo.org/grass/wiki/Release/Schedule#Externalreleaseschedules) (at Trac wiki under Release): The Release Schedule document describes schedule, branching, release maintenance, and numbering. The numbering is changed, specifically the use of odd version numbers for development.
```
Comment (neteler) [inline doc/development/rfc/version_numbering.md:170]: Where shall we maintain this schedule in future?

Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md:170]: Python has release schedules for each minor version as a PEP. Our equivalent would be then a living RFC updated as needed. (The RFC updates are not resolved yet besides the fact that RFC is a misnomer.)

Note: Given the description, the link is meant for the whole schedule, not just the external schedules.

File: doc/development/rfc/version_numbering.md (hunk 17 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md:12]: I'm open to it.

I think the idea for micro was to emphasize that minor is not that small. Patch also has a different meaning like diff or commit although that seems to be diminishing given the spread of PRs/MRs. In the old days, we also had plenty of new features in the "patch" releases than just patches, so micro just stays neutral in that regard.

Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md:12]: Keeping micro for now.

File: doc/development/rfc/version_numbering.md (hunk 18 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md:169]: I mentioned this in #2815.

File: doc/development/rfc/version_numbering.md (hunk 19 above)
Comment (wenzeslaus) [reply doc/development/rfc/version_numbering.md]: Link fixed. Leaving the schedule question for later.

**Comment by wenzeslaus** [comment]
I also implemented the minimal changes required by this RFC. Updating version numbers on main is now the same as on release branch, so the script which was capturing the intricacies of even-odd version distinction is now simpler.

As a consequence, the version on the main branch will no longer be in format X.Y.dev and X.Ydev, but always X.Y.Zdev like the release branches. In other words, with this RFC, there is distinction between even and odd in terms of meaning, there is no distinction in numbering different development branches.

The main branch is simply the upcoming release and the (non-described) branching procedure updates only the version on main, while the branch keeps the original version from main. (Currently both need to be updated.)

The above is a example how this RFC simplifies the branching and release procedure which is an important aspect of this RFC.

The RFC suggests an additional change to the dev labeling (among other things), X.Y.Zdev should become X.Y.Z-dev because that's Semantic Versioning. However, I suggest to implement that separately, perhaps after branching or releasing 8.3.

**Comment by wenzeslaus** [comment]
I guess a number is needed for this RFC. How does that work?

**Comment by wenzeslaus** [comment]
Is accepted as is, it fixes #2335.

File: utils/update_version.md [hunk 20]
```diff
@@ -54,20 +54,22 @@ echo $VERSION
 
 ### Updating Minor Version
 
-Let's say we are at development-only version 3.1.dev and just created
+Let's say were are at the main branch, version 3.2.0dev, and just created
```
Comment (wenzeslaus) [inline utils/update_version.md:57]:
```suggestion
Let's say we are at the main branch, version 3.2.0dev, and just created
```

**Review by hellik — APPROVED** [review]

**Comment by jmckenna** [comment]
Nice addition, +1.  (by the way, here was my own definition of versioning, for the MapServer users : https://github.com/MapServer/MapServer/blob/main/SECURITY.md#version-numbering-explained )

**Comment by neteler** [comment]
As soon as the PSC adopts this RFC (voting is ongoing in https://lists.osgeo.org/pipermail/grass-psc/), the state will be changed to "Approved".
